internal/storage/image.go

```diff
@@ -420,6 +438,209 @@ func (svc *imageService) imageStatus(systemContext *types.SystemContext, unstabl
+	}
+
+	manifestBytes, err := copy.Image(ctx, policyContext, destRef, srcRef, &copy.Options{
+		SourceCtx:      sys,
+		DestinationCtx: sys,
+	})
+	if err != nil {
+		var e errcode.Error
+		if as := errors.As(err, &e); as && e.ErrorCode().Descriptor().HTTPStatusCode == http.StatusUnauthorized {
+			// We cannot support private registries right now, but may with future support from the CRI.
+			log.Errorf(ctx, "Verifying policies against private registries it not supported yet: %v", err)
+			return "", nil
+		}
+
+		var policyErr signature.PolicyRequirementError
+		if errors.As(err, &policyErr) {
+			return "", fmt.Errorf("%w: %w", crierrors.ErrSignatureValidationFailed, err)
+		}
+
+		return "", fmt.Errorf("pulling %q to %q: %w", transports.ImageName(srcRef), transports.ImageName(destRef), err)
+	}
+	return digest.FromBytes(manifestBytes), nil
+}
+
+func (svc *imageService) IsRunningImageAllowed(ctx context.Context, systemContext *types.SystemContext, configImage, userSpecifiedImage string) error {
+	policy, err := signature.DefaultPolicy(systemContext)
+	if err != nil {
+		return fmt.Errorf("get default policy: %w", err)
+	}
+
+	policyContext, err := signature.NewPolicyContext(policy)
+	if err != nil {
+		return fmt.Errorf("create policy context: %w", err)
+	}
+
+	defer func() {
+		if err := policyContext.Destroy(); err != nil {
+			log.Errorf(ctx, "Unable to destroy policy: %v", err)
+		}
+	}()
+
+	var localManifestDigest digest.Digest
+	// On fresh pulls retrieve the full image and digest as configImage
+	if parsedConfigImage, err := reference.Parse(configImage); err == nil {
+		if canonical, isDigested := parsedConfigImage.(reference.Canonical); isDigested {
+			localManifestDigest = canonical.Digest()
+		}
+	}
+
+	var imageParsed reference.Named
+	// Check if the configImage is a local image ID if the image is already pulled
+	if localManifestDigest.String() == "" {
+		img, err := svc.GetStore().Image(configImage)
+		if err != nil && !errors.Is(err, storage.ErrImageUnknown) {
+			return fmt.Errorf("error retrieving image %s: %w", configImage, err)
+		}
+		if img != nil {
+			imageParsed, err = reference.ParseNormalizedNamed(img.Names[0])
+			if err != nil {
+				return fmt.Errorf("failed to parse image name %s into named reference: %w", img.Names[0], err)
+			}
+			localManifestDigest = img.Digest
+		}
+	}
+
+	if imageParsed == nil {
+		imageParsed, err = reference.ParseNormalizedNamed(configImage)
+		if err != nil {
+			return fmt.Errorf("parsing config image %+v : %w", configImage, err)
+		}
+		imageParsed = reference.TagNameOnly(imageParsed)
+	}
+
+	// Tagged references are mainly used by tools like crictl or critest
+	if localManifestDigest.String() == "" {
+		if parsedRef, err := reference.ParseNormalizedNamed(configImage); err == nil {
+			ref, err := svc.pullLocalManifestDigest(ctx, systemContext, policyContext, reference.TagNameOnly(parsedRef))
```

Comment:
⚠️ 

Why is an “is running allowed” predicate triggering a pull? That’s _extremely_ surprising.

(And without credentials it can’t be assumed to succeed, anyway.)